### PR TITLE
Ember guide

### PR DESCRIPTION
**Pull request by danmcclain**

### CODE REVIEW

File: ember.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
+# JavaScript Style Guide
+
+## Table Of Contents
+
+## General
+
+### Create local version of Ember.\* and DS.\*
+
+Future versions of Ember will be released as ES6 modules, so we'll be
```
Comment (rwjblue) [inline ember.md]: ES2015 :trollface:

File: ember.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+able to import `Ember.computed` directly as `computed`
```
Comment (rwjblue) [inline ember.md]: `import computed from 'ember-computed';` will be what it is (at least at this point).

File: ember.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  firstName: DS.attr('string'),
+  lastName: DS.attr('string'),
+
+  fullName: Ember.computed('firstName', 'lastName', function() {
+    // Code
+  })
+});
+```
+
+### Define default values first
```
Comment (rwjblue) [inline ember.md]: These "define" items should be in a "structuring your modules" section or something...

File: ember.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  })
+});
+
+// Bad
+
+export default DS.Model.extend({
+  firstName: DS.attr('string'),
+  lastName: DS.attr('string'),
+
+  fullName: Ember.computed('firstName', 'lastName', function() {
```
Comment (bcardarella) [inline ember.md]: I would even go so far as to make this the prototype extended form, or maybe show both

File: ember.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+Multi-line computed properties should follow your single line computeds.
+
+### Define actions last
+
+Define your route/component/controller's action last, to provide a
+common place that actions can be found.
+
+### Avoid overwriting init
+
+Unless you want to change an object's `init` function, perform actions
```
Comment (bcardarella) [inline ember.md]: you can probably hyperlink to my article about this http://reefpoints.dockyard.com/2014/04/28/dont-override-init.html

File: ember.md (hunk 4 above)
Comment (danmcclain) [inline ember.md]: :+1: Will update

File: ember.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+### Do not use ObjectController or ArrayController
+
+ObjectController is presently deprecated, and ArrayController will be as
+well. Controllers are not going to be used in Ember 2.0, so by using
+`Controller`, you will make it easier to migrate to 2.0.
+
+### Alias your model
+
+It provides a cleaner code to name your model `user` if it is a user. It
```
Comment (bcardarella) [inline ember.md]: is this really a good idea? If I am someone that is jumping into a new code base I know if I see `model` what the purpose is. If I see `user` I would have to figure out what the purpose is

File: ember.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+export default Controller.extend({
+  user: alias('model')
+});
+```
+
+## Templates
+
+### Do not use partials
+
+Always use components. Parials share scope with the parent view, use
```
Comment (bcardarella) [inline ember.md]: Is this really true? If we have a particularly complex page would it be best to break it up into partials? In some cases it would be overkill to split into components. For example, some landing pages are very complex and very focused on content. Content sections don't necessitate components.

File: ember.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+### Use block syntax
+
+Use block syntax instead of `in` syntax with block helpers
+
+```hbs
+{{! Good }}
+{{#each posts as |post|}}
+
+{{! Bad }}
+{{#each post in posts}}
```
Comment (bcardarella) [inline ember.md]: @rwjblue isn't the 2nd form deprecated?

File: ember.md (hunk 7 above)
Comment (rwjblue) [inline ember.md]: @bcardarella - Partials sharing parent scope are likely to be killed in Ember 2.0

File: ember.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+```
+
+### Use components in `{{#each}}` blocks
+
```
Comment (bcardarella) [inline ember.md]: Can we show a code example of this?

File: ember.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+Contents of your each blocks should be a single line, use components
+when more than one line is needed.
+
+## Routes
+
+### Perform all async actions required for the page to load in `model` hooks
+
+The model hooks are async hooks, and will wait for any promises returned
```
Comment (bcardarella) [inline ember.md]: I don't know if I agree with this. I think in most cases it will be true. However, the page rendering is blocked until the model promise resolves. In some cases that data could represent a trivial section of the final page and you can allow it to load async after rendering is complete so as to not block the rest of the page rendering

File: ember.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+to resolve
+
+## Ember Data
+
+### Be explicit with Ember Data attribute types
+
+Even though Ember Data can be used without using explicit types in
```
Comment (bcardarella) [inline ember.md]: how do we feel about attribute annotations? /cc @rwjblue

File: ember.md (hunk 7 above)
Comment (bcardarella) [inline ember.md]: ok, thanks for calrifying

File: ember.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+### Avoid overwriting init
+
+Unless you want to change an object's `init` function, perform actions
+by hooking into the object's `init` hook via `on`. This prevents you
+from forgetting to call `_super`
+
+### Use Pods structure
+
+Store local components within their pod, global components in the
```
Comment (bcardarella) [inline ember.md]: we should probably include the command line options for ember-cli to generate this structure

File: ember.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    model.js
+    serializer.js
+```
+
+## Controllers
+
+### Define query params first
+
+For consistancy and ease of discover, list your query params first in
+yout contoller. These should be listed above default values.
```
Comment (bcardarella) [inline ember.md]: "your"

File: ember.md (hunk 10 above)
Comment (danmcclain) [inline ember.md]: Note the "required for the page to load", I agree with you on your point. If your form has a `<select>` that is bound to a model that will be retrieved async, you should do that in your model, but comments would not be "required" for the page to load and should be retrieve along side. Will update with more explanation

File: ember.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+## Organizing your modules
+
+### Define default values first
+
+Define your object's default values first.
+
+### Define single line computed properties second
+
+Define single line computed properties (`thing: alias('myThing')`)
+after default values
```
Comment (bcardarella) [inline ember.md]: I'm not entirely following what this mean

File: ember.md [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+
+### Define multi-line computed properties third
+
+Multi-line computed properties should follow your single line computeds.
```
Comment (bcardarella) [inline ember.md]: "computeds" => "CPs" or "computed properties"

File: ember.md (hunk 14 above)
Comment (danmcclain) [inline ember.md]: I'll add an example at the top of the this section

File: ember.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+`computed.alias` or `computed.bool`, should be set to `alias` and
+`bool`, respectively
+
+```javascript
+// Good
+
+var computed = Ember.computed;
+var alias = computed.alias;
+var attr = DS.attr;
+var Model = DS.Model;
```
Comment (bcardarella) [inline ember.md]: Where is the line in the sand on how many functions are assigned to local vars? I am all for this with `get/set` and the ones here seem like good ideas. But in some cases it can feel like overkill. I would prefer to stay away from the fuzzy "it depends" area and if we believe that it is best to assign to local vars we should be consistent and do that for all (even if I think it feels like overkill)

File: ember.md (hunk 10 above)
Comment (danmcclain) [inline ember.md]: @bcardarella See this updated section

File: ember.md (hunk 9 above)
Comment (danmcclain) [inline ember.md]: @bcardarella See example below

File: ember.md (hunk 10 above)
Comment (bcardarella) [inline ember.md]: :+1:

File: ember.md (hunk 9 above)
Comment (bcardarella) [inline ember.md]: looks good, I assume the reasoning behind this is to provide something isolated that can be unit tested? If that's the case we should provide that reasoning here

File: ember.md (hunk 16 above)
Comment (rwjblue) [inline ember.md]: I agree.  IMHO, the simplest rule is: do NOT grab anything off of `Ember` or `DS` outside of a top-level variable declaration.  These will all be replaced by direct imports in the future, and keeping them as module level variables makes that future refactoring to use module imports simpler.

File: ember.md (hunk 6 above)
Comment (rwjblue) [inline ember.md]: I do think it is a good idea to use the semantic model name rather than `model`.  `model` is only useful for top level routable components, as soon as you have to pass it down it should be named something useful. Why not just name it something useful/descriptive everywhere?

File: ember.md [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+## Table Of Contents
+
+## General
+
+### Create local version of Ember.\* and DS.\*
+
+Future versions of Ember will be released as ES2015 modules, so we'll be
+able to import `Ember.computed` directly as `computed`. This includes
+`computed.alias` or `computed.bool`, should be set to `alias` and
+`bool`, respectively. Do not use extend prototype syntax
```
Comment (brzpegasus) [inline ember.md]: So... same goes with `Ember.on`, I imagine? `var on = Ember.on;`

What about `Ember.Object`, `Ember.Controller`, and such? Do we need to create local vars for those too (hopefully not)?

File: ember.md (hunk 6 above)
Comment (bcardarella) [inline ember.md]: I just think it leads to an increase in the WTFs/hour. Tracking aliases are just one more thing you have to include in your mental model of each component. Whereas `model`, despite it being a generic name, everyone knows what it is for without having to inspection

File: ember.md (hunk 17 above)
Comment (danmcclain) [inline ember.md]: @brzpegasus Those all fall under Ember.*. In the future, you'll be have to do the following:

```js
import Component from 'ember/component';
```

@rwjblue Can you confirm?

File: ember.md (hunk 16 above)
Comment (brzpegasus) [inline ember.md]: It does feel overkill once you start getting into `var Model = DS.Model`, especially since it only occurs once throughout the file.

File: ember.md (hunk 8 above)
Comment (rwjblue) [inline ember.md]: `{{each post in posts}}` is (or will be) deprecated soon, and `{{each posts}}` is already deprecated.

File: ember.md (hunk 9 above)
Comment (rwjblue) [inline ember.md]: Agreed, a small blurb about why makes sense.

File: ember.md (hunk 6 above)
Comment (danmcclain) [inline ember.md]: I would call a component this way `{{post-summary post=model}}` vs `{{post-summary model=model}}`.

File: ember.md (hunk 16 above)
Comment (bcardarella) [inline ember.md]: I wonder if it would make sense to look at this from a footprint point of view. In same cases declaring a local variable will result in more characters than just using the function off the parent object. However, at some point using that function multiple times will eventually cost more in characters than declaring a local variable. Would that make for a good guide on when to declare local vars?

File: ember.md (hunk 6 above)
Comment (bcardarella) [inline ember.md]: @danmcclain I agree with that but that isn't what this section describes. I read this as taking a routable controller/component that is having `model` auto-assigned then aliasing it to another property

File: ember.md (hunk 17 above)
Comment (rwjblue) [inline ember.md]: Yea, but I think it will be (slight tweak in the module path):

```
import Component from 'ember-component';
```

@brzpegasus - Basically yes, the idea being that when we can import them directly, it is simple to replace the `var` declarations at the top of the file with `import` statements without having to parse all usages out of the code manually.

File: ember.md (hunk 16 above)
Comment (rwjblue) [inline ember.md]: `Ember.Object` cannot be minified further than `E.Object` (since you are chaining), but `var EmberObject = Ember.Object` can minify `EmberObject` down to a single character for all usage after the var declaration.  Also, once gzip'ed the shared strings actually compress very well (in Ember itself we try to organize imports in the same order so larger chunks are the same and can be completely replaced by gzip).

File: ember.md (hunk 6 above)
Comment (rwjblue) [inline ember.md]: When we have routable components, we will basically have an `attr` hook that sets the properties of the top level routable component. The current proposal is something like:

```javascript
export default Route.extend({
  attrs () {
    return {
      model: this.model()
    }
  }
});
```

The above `attrs` hook will pass through `RSVP.hash`, so that `SomeComponent.create({ model: /* whatever was returned by model hook */ })` will be called.

---

So given the above direction, this would be done in 2.0/routable component world without an alias:

```javascript
export default Route.extend({
  attrs () {
    return {
      user: this.model()
    }
  }
});
```

---

**tldr;** The `attrs` hook is essentially just like invoking your component from a template with those hash arguments.  You would generally not do `{{some-thing model=model}}` (because in `SomeThingComponent` `model` is not a useful name), you would do `{{some-thing thing=model}}` (or whatever name made sense in your problem space).

Comment (bcardarella) [inline ember.md]: ah, so `model` will no longer be a convention. Ok, I can accept that

Comment (rwjblue) [inline ember.md]: @bcardarella - It may be the default implementation (mostly for compat reasons), but I think that we will migrate to using a "domain word" instead of `model` in the guides once this all lands.

File: ember.md (hunk 11 above)
Comment (rwjblue) [inline ember.md]: We should hold off on annotations until they are formally proposed to TC39.  I believe that this is actively underway though, so it should not be too long...

**Comment by danmcclain** [comment]
Any objections to merging this? All comments have been addressed. I'll let this simmer for another day before merging to allow for any additional comments
